File: rust/src/medium/diamond.rs
pub fn get_diamond(c: char) -> Vec<String> {
    let range = 'A'..=c;

    let letters = range.clone().chain(range.clone().rev().skip(1));
    let length = range.count();

    let row_seq = (0..length).chain((0..length).rev().skip(1)).zip(letters);
    let col_seq = (0..length).rev().chain((0..length).skip(1));

    row_seq
        .map(|(row, letter)| {
            col_seq
                .clone()
                .map(|col| if col == row { letter } else { ' ' })
                .collect()
        })
        .collect()
}

#[test]
fn test_d() {
    #[rustfmt::skip]
    assert_eq!(
        get_diamond('D'),
        vec![
            "   A   ",
            "  B B  ",
            " C   C ",
            "D     D",
            " C   C ",
            "  B B  ",
            "   A   ",
        ]
    );
}

File: elixir/lib/medium/diamond.ex
defmodule Diamond do
  @doc """
  Given a letter, it prints a diamond starting with 'A',
  with the supplied letter at the widest point.
  """
  @spec build_shape(char) :: String.t()
  def build_shape(letter) do
    letters = letters(letter)
    center = Enum.count(?A..letter)

    letters
    |> Enum.zip(rows(center))
    |> Enum.map(fn {letter, row} -> render_row(cols(center), letter, row) end)
    |> Enum.join()
  end

  defp render_row(cols, letter, row) do
    chars = Enum.map(cols, fn col -> if row == col, do: letter, else: ' ' end)

    List.to_string(chars) <> "\n"
  end

  defp letters(?A), do: [?A]
  defp letters(letter), do: Enum.concat(?A..(letter - 1), letter..?A)

  defp cols(1), do: [0]
  defp cols(num), do: Enum.concat(0..(num - 2), (num - 1)..0)

  defp rows(1), do: [0]
  defp rows(num), do: Enum.concat((num - 1)..0, 1..(num - 1))
end
